File: server/src/routes/payments.js
import Stripe from 'stripe';
import express from 'express';

// Initialize Stripe with your secret key
const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || 'sk_test_...', {
    apiVersion: '2023-10-16',
});

const router = express.Router();

/**
 * Create a Stripe Checkout Session for subscription
 * POST /api/payments/create-checkout-session
 */
router.post('/create-checkout-session', async (req, res) => {
    try {
        const { priceId, userId, email, tier } = req.body;

        if (!priceId || !userId || !email) {
            return res.status(400).json({ error: 'Missing required fields' });
        }

        // Create Checkout Session
        const session = await stripe.checkout.sessions.create({
            mode: 'subscription',
            payment_method_types: ['card'],
            line_items: [
                {
                    price: priceId, // Stripe Price ID (e.g., price_1234...)
                    quantity: 1,
                },
            ],
            customer_email: email,
            metadata: {
                userId,
                tier,
            },
            success_url: `${process.env.FRONTEND_URL}/dashboard?session_id={CHECKOUT_SESSION_ID}`,
            cancel_url: `${process.env.FRONTEND_URL}/pricing`,
            subscription_data: {
                metadata: {
                    userId,
                    tier,
                },
            },
        });

        res.json({ sessionId: session.id, url: session.url });
    } catch (error) {
        console.error('Stripe checkout error:', error);
        res.status(500).json({ error: 'Failed to create checkout session' });
    }
});

/**
 * Create a Stripe Customer Portal Session for managing subscription
 * POST /api/payments/create-portal-session
 */
router.post('/create-portal-session', async (req, res) => {
    try {
        const { customerId } = req.body;

        if (!customerId) {
            return res.status(400).json({ error: 'Customer ID required' });
        }

        const session = await stripe.billingPortal.sessions.create({
            customer: customerId,
            return_url: `${process.env.FRONTEND_URL}/dashboard`,
        });

        res.json({ url: session.url });
    } catch (error) {
        console.error('Stripe portal error:', error);
        res.status(500).json({ error: 'Failed to create portal session' });
    }
});

/**
 * Webhook handler for Stripe events
 * POST /api/payments/webhook
 */
router.post('/webhook', express.raw({ type: 'application/json' }), async (req, res) => {
    const sig = req.headers['stripe-signature'];

    let event;

    try {
        event = stripe.webhooks.constructEvent(
            req.body,
            sig,
            process.env.STRIPE_WEBHOOK_SECRET
        );
    } catch (err) {
        console.error('Webhook signature verification failed:', err.message);
        return res.status(400).send(`Webhook Error: ${err.message}`);
    }

    // Handle the event
    switch (event.type) {
        case 'checkout.session.completed': {
            const session = event.data.object;
            console.log('Checkout session completed:', session.id);

            // TODO: Update user subscription in database
            const { userId, tier } = session.metadata;
            // await updateUserSubscription(userId, {
            //   tier,
            //   status: 'active',
            //   stripeCustomerId: session.customer,
            //   stripeSubscriptionId: session.subscription,
            // });

            break;
        }

        case 'customer.subscription.updated': {
            const subscription = event.data.object;
            console.log('Subscription updated:', subscription.id);

            // TODO: Update subscription status
            break;
        }

        case 'customer.subscription.deleted': {
            const subscription = event.data.object;
            console.log('Subscription cancelled:', subscription.id);

            // TODO: Downgrade user to free tier
            // const userId = subscription.metadata.userId;
            // await updateUserSubscription(userId, { tier: 'free', status: 'canceled' });

            break;
        }

        case 'invoice.payment_succeeded': {
            const invoice = event.data.object;
            console.log('Payment succeeded:', invoice.id);
            break;
        }

        case 'invoice.payment_failed': {
            const invoice = event.data.object;
            console.log('Payment failed:', invoice.id);

            // TODO: Send email notification to user
            break;
        }

        default:
            console.log(`Unhandled event type: ${event.type}`);
    }

    res.json({ received: true });
});

/**
 * Get subscription details
 * GET /api/payments/subscription/:subscriptionId
 */
router.get('/subscription/:subscriptionId', async (req, res) => {
    try {
        const { subscriptionId } = req.params;

        const subscription = await stripe.subscriptions.retrieve(subscriptionId);

        res.json({
            status: subscription.status,
            currentPeriodEnd: new Date(subscription.current_period_end * 1000),
            cancelAtPeriodEnd: subscription.cancel_at_period_end,
            plan: subscription.items.data[0].price.id,
        });
    } catch (error) {
        console.error('Failed to retrieve subscription:', error);
        res.status(500).json({ error: 'Failed to retrieve subscription' });
    }
});

/**
 * Cancel subscription
 * POST /api/payments/cancel-subscription
 */
router.post('/cancel-subscription', async (req, res) => {
    try {
        const { subscriptionId } = req.body;

        if (!subscriptionId) {
            return res.status(400).json({ error: 'Subscription ID required' });
        }

        // Cancel at period end (don't immediately cancel)
        const subscription = await stripe.subscriptions.update(subscriptionId, {
            cancel_at_period_end: true,
        });

        res.json({
            status: subscription.status,
            cancelAtPeriodEnd: subscription.cancel_at_period_end,
            currentPeriodEnd: new Date(subscription.current_period_end * 1000),
        });
    } catch (error) {
        console.error('Failed to cancel subscription:', error);
        res.status(500).json({ error: 'Failed to cancel subscription' });
    }
});

export default router;

File: backend-python/app/api/v1/routes/payments.py
"""
Payment Routes - Stripe subscription management
"""

from fastapi import APIRouter, Depends, HTTPException, status, Request, Header
from motor.motor_asyncio import AsyncIOMotorDatabase
from pydantic import BaseModel

from app.core.config import settings
from app.core.database import get_database
from app.models.user import User
from app.services.user_service import UserService
from app.services.payment_service import PaymentService
from app.middleware.auth import get_current_user

router = APIRouter(prefix="/payments", tags=["payments"])


# Request models
class CreateCheckoutSession(BaseModel):
    tier: str  # "pro" or "enterprise"
    success_url: str
    cancel_url: str


class UpdateSubscription(BaseModel):
    tier: str  # "pro" or "enterprise"


# Dependency
async def get_user_service(db: AsyncIOMotorDatabase = Depends(get_database)) -> UserService:
    return UserService(db)


@router.post("/create-checkout-session")
async def create_checkout_session(
    request: CreateCheckoutSession,
    current_user: User = Depends(get_current_user),
    user_service: UserService = Depends(get_user_service)
):
    """
    Create a Stripe Checkout session for subscription
    
    Steps:
    1. Create Stripe customer if doesn't exist
    2. Create checkout session
    3. Return checkout URL
    """
    # Validate tier
    if request.tier not in ["pro", "enterprise"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid subscription tier"
        )
    
    # Get price ID
    price_id = (
        settings.STRIPE_PRO_PRICE_ID 
        if request.tier == "pro" 
        else settings.STRIPE_ENTERPRISE_PRICE_ID
    )
    
    if not price_id:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Stripe price ID not configured"
        )
    
    try:
        # Create customer if doesn't exist
        customer_id = current_user.subscription.stripe_customer_id
        if not customer_id:
            customer_id = await PaymentService.create_customer(
                email=current_user.email,
                name=current_user.full_name,
                user_id=str(current_user.id)
            )
            # Update user with customer ID
            current_user.subscription.stripe_customer_id = customer_id
            await user_service.update_subscription(
                str(current_user.id),
                current_user.subscription
            )
        
        # Create checkout session
        checkout_url = await PaymentService.create_checkout_session(
            customer_id=customer_id,
            price_id=price_id,
            success_url=request.success_url,
            cancel_url=request.cancel_url,
            tier=request.tier
        )
        
        return {"checkout_url": checkout_url}
    
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/create-subscription")
async def create_subscription(
    tier: str,
    current_user: User = Depends(get_current_user),
    user_service: UserService = Depends(get_user_service)
):
    """
    Create subscription directly (returns client secret for frontend payment)
    """
    if tier not in ["pro", "enterprise"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid subscription tier"
        )
    
    price_id = (
        settings.STRIPE_PRO_PRICE_ID 
        if tier == "pro" 
        else settings.STRIPE_ENTERPRISE_PRICE_ID
    )
    
    try:
        # Create customer if doesn't exist
        customer_id = current_user.subscription.stripe_customer_id
        if not customer_id:
            customer_id = await PaymentService.create_customer(
                email=current_user.email,
                name=current_user.full_name,
                user_id=str(current_user.id)
            )
        
        # Create subscription
        result = await PaymentService.create_subscription(
            customer_id=customer_id,
            price_id=price_id
        )
        
        return result
    
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/cancel-subscription")
async def cancel_subscription(
    current_user: User = Depends(get_current_user),
    user_service: UserService = Depends(get_user_service)
):
    """Cancel current subscription"""
    subscription_id = current_user.subscription.stripe_subscription_id
    
    if not subscription_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No active subscription"
        )
    
    try:
        await PaymentService.cancel_subscription(subscription_id)
        
        # Update user subscription to free
        current_user.subscription.tier = "free"
        current_user.subscription.status = "canceled"
        current_user.subscription.stripe_subscription_id = None
        await user_service.update_subscription(
            str(current_user.id),
            current_user.subscription
        )
        
        return {"message": "Subscription canceled successfully"}
    
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/update-subscription")
async def update_subscription(
    request: UpdateSubscription,
    current_user: User = Depends(get_current_user),
    user_service: UserService = Depends(get_user_service)
):
    """Update subscription to different tier"""
    if request.tier not in ["pro", "enterprise"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid subscription tier"
        )
    
    subscription_id = current_user.subscription.stripe_subscription_id
    if not subscription_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No active subscription to update"
        )
    
    price_id = (
        settings.STRIPE_PRO_PRICE_ID 
        if request.tier == "pro" 
        else settings.STRIPE_ENTERPRISE_PRICE_ID
    )
    
    try:
        result = await PaymentService.update_subscription(
            subscription_id=subscription_id,
            new_price_id=price_id
        )
        
        # Update user subscription
        current_user.subscription.tier = request.tier
        current_user.subscription.status = result["status"]
        current_user.subscription.current_period_end = result["current_period_end"]
        await user_service.update_subscription(
            str(current_user.id),
            current_user.subscription
        )
        
        return {"message": "Subscription updated successfully"}
    
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get("/customer-portal")
async def customer_portal(
    current_user: User = Depends(get_current_user)
):
    """Get Stripe customer portal URL"""
    customer_id = current_user.subscription.stripe_customer_id
    
    if not customer_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No Stripe customer found"
        )
    
    try:
        portal_url = await PaymentService.create_customer_portal_session(
            customer_id=customer_id,
            return_url=f"{settings.FRONTEND_URL}/dashboard"
        )
        
        return {"portal_url": portal_url}
    
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/webhook")
async def stripe_webhook(
    request: Request,
    stripe_signature: str = Header(None, alias="stripe-signature"),
    user_service: UserService = Depends(get_user_service)
):
    """
    Handle Stripe webhook events
    
    Events handled:
    - checkout.session.completed
    - customer.subscription.updated
    - customer.subscription.deleted
    - invoice.payment_succeeded
    - invoice.payment_failed
    """
    # Get raw body
    payload = await request.body()
    
    # Verify webhook signature
    event = PaymentService.verify_webhook_signature(
        payload=payload,
        signature=stripe_signature,
        webhook_secret=settings.STRIPE_WEBHOOK_SECRET
    )
    
    if not event:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid webhook signature"
        )
    
    event_type = event["type"]
    event_data = event["data"]
    
    try:
        if event_type == "checkout.session.completed":
            # Payment successful, subscription created
            session = event_data["object"]
            customer_id = session["customer"]
            subscription_id = session["subscription"]
            
            # Find user by customer ID
            db = await get_database()
            user_data = await db.users.find_one(
                {"subscription.stripe_customer_id": customer_id}
            )
            
            if user_data:
                user = User(**user_data)
                # Update subscription will be handled by subscription.updated event
        
        elif event_type == "customer.subscription.updated":
            # Subscription status changed
            subscription_data = await PaymentService.handle_subscription_updated(event_data)
            
            # Find user by customer ID
            db = await get_database()
            user_data = await db.users.find_one(
                {"subscription.stripe_customer_id": subscription_data.stripe_customer_id}
            )
            
            if user_data:
                user = User(**user_data)
                await user_service.update_subscription(
                    str(user.id),
                    subscription_data
                )
        
        elif event_type == "customer.subscription.deleted":
            # Subscription canceled
            subscription_data = await PaymentService.handle_subscription_deleted(event_data)
            
            # Find user by customer ID
            db = await get_database()
            user_data = await db.users.find_one(
                {"subscription.stripe_customer_id": subscription_data.stripe_customer_id}
            )
            
            if user_data:
                user = User(**user_data)
                await user_service.update_subscription(
                    str(user.id),
                    subscription_data
                )
        
        elif event_type == "invoice.payment_failed":
            # Payment failed - notify user
            invoice = event_data["object"]
            customer_id = invoice["customer"]
            # TODO: Send email notification
        
        return {"status": "success"}
    
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Webhook processing failed: {str(e)}"
        )


@router.get("/subscription-status")
async def get_subscription_status(
    current_user: User = Depends(get_current_user)
):
    """Get current subscription status"""
    subscription = current_user.subscription
    
    result = {
        "tier": subscription.tier,
        "status": subscription.status,
        "customer_id": subscription.stripe_customer_id,
        "subscription_id": subscription.stripe_subscription_id,
        "current_period_end": subscription.current_period_end
    }
    
    # Get live status from Stripe if subscription exists
    if subscription.stripe_subscription_id:
        stripe_data = await PaymentService.get_subscription(
            subscription.stripe_subscription_id
        )
        if stripe_data:
            result["stripe_status"] = stripe_data["status"]
            result["cancel_at_period_end"] = stripe_data["cancel_at_period_end"]
    
    return result
